### ISSUE TITLE
aursync -u installs packages immediately

### ISSUE THREAD

**Issue by clawoflight** [AladW/aurutils#152]
I assume that this is a feature, but it would be nice to be able to turn it off.

I would like to use `aursync -u --no-install` or similar to update the packages in my local repo without immediately having pacman install them.

Is this planned, or am I missing something?

**Comment by clawoflight** [comment]
After reading the source code of `aursync` and `aurbuild`, it seems to me that this is not the intended behaviour. Do you have any idea why I this could be happening?

**Comment by clawoflight** [comment]
O, I missed it. That happens in [Line 49](https://github.com/AladW/aurutils/blob/master/bin/aurbuild#L49) of `aurbuild`.

Is there a reason why that would need to happen?
Could it be disabled by default or through a flag?

**Comment by runical** [comment]
That line updates the repository and updates packages. If the update is installed on the computer, the package will be installed. This is needed when building packages which depend on packages built in the same session.

If you want to disable it, you could add something like `--ignore=$PKGNAMES` to the pacman call, but that will disable building packages in a chain.

**Comment by clawoflight** [comment]
Thanks, that makes sense :)

So there's no straight-forward, clean way to achieve what I want, right?

**Comment by runical** [comment]
Not really as far as I can see, but perhaps Alad or someone else has an idea. 

The quickest way to get the behaviour you want would be to build in a chroot. In that case the local database is not updated, so no updates. You also get a clean package which fails to build in case some dependency is missing which is an added bonus (for me at least).

**Comment by AladW** [comment]
The only thing that installs packages in aurbuild is `makepkg --sync`. It does so by checking needed dependencies with `pacman -T`, then installing the remainders with `pacman -S --asdeps`. As the default options (for aursync) also include `makepkg --rmdeps`, the dependencies are removed after building the package.

`pacman -Syu` is used to prevent building against outdated libraries. Per the above, if a package is already installed (because it is needed by a different package, or you've installed it explicitely), makepkg's `pacman -S` wouldn't update it.

In other words, unless you're using `aursync --keep` which unsets `makepkg --rmdeps`, installed a package explicitely, or prompted `n` on a query to remove dependencies, packages are removed after the build.

If you have a reproducible case that shows otherwise, please checkout from git and provide a `bash -x` log as indicated in the Troubleshooting section of the README.

**Comment by AladW** [comment]
Any more feedback on this?

**Comment by clawoflight** [comment]
Excuse me, I was overwhelmed by my exams and forgot to reply.

What I was trying to achieve was to be able to run `aursync -u` without 
getting the new versions of AUR packages on my system, thereby only updating 
my local repo.

I haven't tried that yet because of the exams, but building in a chroot like 
@runical mentioned (I assume `aursync -cu`) should do exactly what I want.

I will try that this week and report back if you want, or you can close this 
now.

**Comment by AladW** [comment]
Oh, I see what you mean now. Yes, in that case building in a chroot is the best approach; perhaps I should rewrite some of the comments in aurbuild, and move them to the man page to make this more clear.

And yes, please let me know how it works out. When testing, I'd suggest the version from git as I've planned a few changes for the next release.

**Comment by clawoflight** [comment]
@AladW I can confirm that `aursync -cu` behaves like I expected.
Thanks a lot for the changes to the man page, they are very helpful :)

I have another question though: When building in the chroot, my `/etc/makepkg.conf` seems to be ignored. Is that intended?

Thanks a lot for your help!

**Comment by AladW** [comment]
It seems that arch-nspawn only copies the host /etc/makepkg.conf (or other specified makepkg.conf) if you do so explicitely with `-M`; however, the DEST variables, MAKEFLAGS and PACKAGER are read by makechrootpkg (aurbuild overrides PKGDEST to move the packages to the repository pool).

What variables in particular are you missing?

**Comment by clawoflight** [comment]
You are so fast :)

I am not _really_ missing anything. I noticed it because I have `sign` set in `BUILDENV` because I signed packages before switching to aurutils, where `-s` is an option.
I am happy as long as MAKEFLAGS and PACKAGER are set, I will simply use `aursync -csu`  from now on.

I just wanted to fully understand everything.

Thanks a lot! I will close this now :I

**Comment by wesbarnett** [comment]
I realize this issue is closed, but I came across this since I had a similar thought as OP. systemd-nspawn has a nice ephemeral feature for btrfs systems that can be useful for this. Here's a one liner I do to upgrade my AUR packages in my local repo without installing them:

    sudo systemd-nspawn --user=$USER --ephemeral --directory=/ --bind=/var/cache/pacman/aur bash -c "yes | aursync -u"

I didn't want to have an arch container that I would need to update and keep on the system. With this the container is temporary, is a snapshot of my (up-to-date) root subvolume, and updates my local repository on the system located at `/var/cache/pacman/aur`. Also I don't need the devtools package for this. Anyways, just wanted to share this for anyone else who comes across it.

Thanks for the work you are doing with this great package.

**Comment by tadfisher** [comment]
Also +1 on installing needed dependencies in the nspawn container without updating packages the host machine. Ideally I could run a systemd unit which updates a local repo without intervention or modifying the host environment.

**Comment by AladW** [comment]
Interesting trick with systemd-nspawn. Regarding the systemd unit, you could use a timer. Please note that the service below uses `--noview`, so you should only use this for packages you maintain yourself or by people you trust.

**/etc/systemd/system/aursync.service**
```
[Unit]
Description=Local repository update
Documentation=man:aursync(1)
Requires=network-online.target
After=network-online.target

[Service]
Type=oneshot
ExecStart=/usr/bin/aursync -cu --repo <yourcustomrepo> --noview
```
**/etc/systemd/system/aursync.timer**
```
cat reflector.timer 
[Unit]
Description=Daily local repository update

[Timer]
OnCalendar=daily
Persistent=true
AccuracySec=12h

[Install]
WantedBy=timers.target
```

**Comment by AladW** [comment]
I found some way to fit in a ```no_sync``` option that simply wouldn't call pacman at all. Clearly this breaks dependency resolution (makepkg can't install dependencies that aren't synced to a pacman db), but it might make sense for smaller sense. Suggestions on what to call the (short) command-line option?

**Comment by clawoflight** [comment]
I'm not sure.

-n for --no-sync, but there are many other --no-* options.
-i for --isolated or something, but that sounds misleading.
-o for --oneshot, but same problem as above

So I don't really have a good idea...

On 02/03/2017 12:24 AM, Alad Wenter wrote:
> I found some way to fit in a ```no_sync``` option that simply
> wouldn't call pacman at all. Clearly this breaks dependency
> resolution (makepkg can't install dependencies that aren't synced to
> a pacman db), but it might make sense for smaller sense. Suggestions
> on what to call the (short) command-line option?
> 

-- 
GPG fingerprint: 871F 1047 7DB3 DDED 5FC4 47B2 26C7 E577 EF96 7808
